// nbit-light-chart@1.0.9/es/packages/light-chart/src/light-chart/util/theme.js
var getTheme = function getTheme() {
  var bgColor = getComputedStyle(document.body).getPropertyValue('--card_bg_color_01');
  var textColor = getComputedStyle(document.body).getPropertyValue('--text_color_01');
  var brandColor = getComputedStyle(document.body).getPropertyValue('--brand_color').replace(/(^\s*)|(\s*$)/g, '');
  var upColor = getComputedStyle(document.body).getPropertyValue('--buy_up_color').replace(/(^\s*)|(\s*$)/g, '');
  var downColor = getComputedStyle(document.body).getPropertyValue('--sell_down_color').replace(/(^\s*)|(\s*$)/g, '');
  var upLightColor = getComputedStyle(document.body).getPropertyValue('--buy_up_color_hover').replace(/(^\s*)|(\s*$)/g, '');
  var downLightColor = getComputedStyle(document.body).getPropertyValue('--sell_down_color_hover').replace(/(^\s*)|(\s*$)/g, '');
  return {
    bgColor: bgColor,
    textColor: textColor,
    brandColor: brandColor,
    upColor: upColor,
    downColor: downColor,
    upLightColor: upLightColor,
    downLightColor: downLightColor
  };
};

export { getTheme };
