// nbit-light-chart@1.0.9/lib/packages/light-chart/src/light-chart/components/dept/dept.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
require('./dept.css.js');
var deptChart = require('./dept-chart.js');
var deptRealChart = require('./dept-real-chart.js');
var theme = require('../../util/theme.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

exports.DeptChartSpecieEnum = void 0;
(function (DeptChartSpecieEnum) {
    DeptChartSpecieEnum["Dept"] = "dept";
    DeptChartSpecieEnum["DeptCurrent"] = "deptCurrent";
})(exports.DeptChartSpecieEnum || (exports.DeptChartSpecieEnum = {}));
const Dept = (props) => {
    const { bgColor, textColor, brandColor, upColor, downColor, upLightColor, downLightColor } = theme.getTheme();
    const [chartLayoutOptions, setChartLayoutOptions] = React.useState({
        background: {
            color: bgColor,
        },
        textColor,
    });
    React.useEffect(() => {
        const _bgColor = theme.getTheme().bgColor;
        const _textColor = theme.getTheme().textColor;
        setChartLayoutOptions({
            background: {
                color: _bgColor,
            },
            textColor: _textColor,
        });
    }, [props.theme]);
    return (React__default["default"].createElement("div", { className: "dept-chart-wrap" },
        React__default["default"].createElement("div", { className: "chart-wrap" }, props.type === exports.DeptChartSpecieEnum.Dept ? (React__default["default"].createElement(deptChart, { chartLayoutOptions: chartLayoutOptions, data: props.deptData, createChart: { brandColor, upColor, downColor, upLightColor, downLightColor } })) : (React__default["default"].createElement(deptRealChart["default"], { chartLayoutOptions: chartLayoutOptions, data: props.deptData, createChart: { brandColor, upColor, downColor, upLightColor, downLightColor } })))));
};
var dept = React.memo(Dept);

exports["default"] = dept;
