// nbit-light-chart@1.0.9/src/light-chart/components/series/series.tsx
import React, { useEffect, useState, memo } from 'react'
import './series.css'
import SeriesChart from './series-chart'
import { getTheme } from '../../util/theme'

export interface SeriesChartData {
  time: number // 13 位时间戳
  value: number
  [key: string]: any
}

export interface PropsType {
  theme: string
  seriesData: Array<SeriesChartData>
}

const Series = (props: PropsType) => {
  const { bgColor, textColor, brandColor } = getTheme()

  const [chartLayoutOptions, setChartLayoutOptions] = useState({
    background: {
      color: bgColor,
    },
    textColor,
  })

  useEffect(() => {
    const _bgColor = getTheme().bgColor
    const _textColor = getTheme().textColor
    setChartLayoutOptions({
      background: {
        color: _bgColor,
      },
      textColor: _textColor,
    })
  }, [props.theme])

  return (
    <div className="series-chart-wrap">
      <div className="chart-wrap">
        <SeriesChart chartLayoutOptions={chartLayoutOptions} data={props.seriesData} createChart={{ brandColor }} />
      </div>
    </div>
  )
}

export default memo(Series)
