// @nbit/chart-h5@1.0.0-beta.54/src/components/dept/index.tsx
import React, { useEffect, useState, memo } from 'react'
import './index.css'
import { getTheme, DeptChartData, DeptChartSpecieEnum } from '@nbit/chart-utils'
import DeptChart from './dept-chart'
import DeptRealChart from './dept-real-chart'

interface PropsType {
  theme: string
  deptData: Array<DeptChartData>
  type?: DeptChartSpecieEnum
  offset: {
    priceOffset: number
    amountOffset: number
  }
  colors: number
}

function Dept(props: PropsType) {
  const {
    bgColor,
    textColor,
    textColor01,
    brandColor,
    upColor,
    downColor,
    upLightColor,
    downLightColor,
    cardBgColor,
    cardBgColor03,
  } = getTheme()
  const [chartLayoutOptions, setChartLayoutOptions] = useState({
    background: {
      color: bgColor,
    },
    textColor,
  })

  useEffect(() => {
    const _bgColor = getTheme().bgColor
    const _textColor = getTheme().textColor
    setChartLayoutOptions({
      background: {
        color: _bgColor,
      },
      textColor: _textColor,
    })
  }, [props.theme])

  const { priceOffset } = props.offset

  return (
    <div className="dept-chart-wrap">
      <div className="dept-chart">
        {props.type === DeptChartSpecieEnum.Dept ? (
          <DeptChart
            chartLayoutOptions={chartLayoutOptions}
            data={props.deptData}
            createChart={{ brandColor, upColor, downColor, upLightColor, downLightColor }}
          />
        ) : (
          <DeptRealChart
            chartLayoutOptions={chartLayoutOptions}
            data={props.deptData}
            theme={props.theme}
            priceOffset={priceOffset}
            colors={props.colors}
            createChart={{
              brandColor,
              upColor,
              downColor,
              upLightColor,
              downLightColor,
              bgColor,
              textColor,
              cardBgColor,
              textColor01,
              cardBgColor03,
            }}
          />
        )}
      </div>
    </div>
  )
}

export default memo(Dept)
