// nbit-light-chart@1.0.9/src/light-chart/components/dept/dept.tsx
import React, { useEffect, useState, memo } from 'react'
import './dept.css'
import DeptChart from './dept-chart'
import DeptRealChart from './dept-real-chart'
import { getTheme } from '../../util/theme'

export interface DeptChartData {
  time: number // 13 位时间戳
  value: number
  [key: string]: any
}

export enum DeptChartSpecieEnum {
  Dept = 'dept',
  DeptCurrent = 'deptCurrent',
}
export interface PropsType {
  theme: string
  deptData: Array<DeptChartData>
  type?: DeptChartSpecieEnum
}

const Dept = (props: PropsType) => {
  const { bgColor, textColor, brandColor, upColor, downColor, upLightColor, downLightColor } = getTheme()
  const [chartLayoutOptions, setChartLayoutOptions] = useState({
    background: {
      color: bgColor,
    },
    textColor,
  })

  useEffect(() => {
    const _bgColor = getTheme().bgColor
    const _textColor = getTheme().textColor
    setChartLayoutOptions({
      background: {
        color: _bgColor,
      },
      textColor: _textColor,
    })
  }, [props.theme])

  return (
    <div className="dept-chart-wrap">
      <div className="chart-wrap">
        {props.type === DeptChartSpecieEnum.Dept ? (
          <DeptChart
            chartLayoutOptions={chartLayoutOptions}
            data={props.deptData}
            createChart={{ brandColor, upColor, downColor, upLightColor, downLightColor }}
          />
        ) : (
          <DeptRealChart
            chartLayoutOptions={chartLayoutOptions}
            data={props.deptData}
            createChart={{ brandColor, upColor, downColor, upLightColor, downLightColor }}
          />
        )}
      </div>
    </div>
  )
}

export default memo(Dept)
