// nbit-light-chart@1.0.9/src/light-chart/components/dept/dept-chart.tsx
import React, { useEffect, useRef } from 'react'
import { ChartOptions, SingleValueData, createChart } from 'lightweight-charts'
import dayjs from 'dayjs'
import { DeptChartData } from './dept'

export interface PropsType {
  data: Array<DeptChartData>
  chartLayoutOptions: any
  createChart: {
    brandColor: string
    upColor: string
    downColor: string
    upLightColor: string
    downLightColor: string
  }
}

const DeptChart = (props: PropsType) => {
  const deptRef = useRef<any>(null)
  const chartRef = useRef<any>(null)

  useEffect(() => {
    const chart = createChart(document.getElementById('chart') as HTMLElement, {
      layout: props.chartLayoutOptions,
      // 时间刻度和价格刻度
      timeScale: {
        visible: false,
      },
      rightPriceScale: {
        visible: false,
      },
      // 布局相关
      crosshair: {
        horzLine: {
          visible: false,
          labelVisible: false,
        },
        vertLine: {
          visible: false,
          labelVisible: false,
        },
      },
      grid: {
        vertLines: {
          visible: false,
        },
        horzLines: {
          visible: false,
        },
      },
    })

    const isUp = props.data[props.data.length - 1].value - (props.data[props.data.length - 2]?.value || 0) > 0
    chartRef.current = chart.addAreaSeries({
      lineColor: isUp ? props.createChart.upColor : props.createChart.downColor,
      topColor: isUp ? props.createChart.upColor : props.createChart.downColor,
      bottomColor: isUp ? props.createChart.upLightColor : props.createChart.downLightColor,
      crosshairMarkerVisible: false,
      priceLineVisible: false,
    })

    chartRef.current.setData(props.data)
    chart.timeScale().fitContent()
  }, [])

  return <div ref={deptRef} className="chart" id="chart"></div>
}

DeptChart.displayName = 'deptChart'

export default DeptChart
